<!DOCTYPE html>
<html>
<head>
<title>synthjs Unit Tests</title>
<script src="../../closure/goog/base.js"></script>
<script src="../../deps.js"></script>
<script>
	goog.require("goog.testing.jsunit");
	goog.require("goog.testing.AsyncTestCase");
	
	goog.require("synthjs.utility.Deferred");
	goog.require("synthjs.utility.WorkerDeferred");
</script>
</head>
<body>
<script>

var asyncTestCase = new goog.testing.AsyncTestCase.createAndInstall();
var D = synthjs.utility.Deferred;

function setUp() {

}

function tearDown() {

}

function testDeferredObject() {
	var cnt = 0;
	var WD = synthjs.utility.WorkerDeferred;
	
	var worker = new Worker('test_echo_worker.js');
	var d = new WD(worker, {val:1}).addCallback(function(e){
		assertEquals(e.val, 1);
		asyncTestCase.continueTesting();
	});
	
	asyncTestCase.waitForAsync();
	d.callback();
};

function testDeferredCallback() {
	var WD = synthjs.utility.WorkerDeferred;
	
	var worker = new Worker('test_echo_worker.js');
	
	var dWait = new D();
	
	var d = new D().addCallback(function(){
			return {val: 'a'};
		}).assocChainDeferred(
//			new D().addCallback(function(){
//				dWait.callback();
//			}).awaitDeferred(dWait)
			new WD(worker, function(e){
				e.add = 'b';
				return e;
			})
		).addCallback(function(e){
			assertEquals(e.add, 'b');
			asyncTestCase.continueTesting();
		});
	
	asyncTestCase.waitForAsync();
	d.callback();
}

/*
 * try-catchではcatchできない。window.onerrorでキャッチ可能
 */
function _testFailure() {
	var cnt = 0;
	var WD = synthjs.utility.WorkerDeferred;
	
	var worker = new Worker('test_echo_worker_failed.js');
	var d = new WD(worker, {val:1}, {
		error: function(){
			return 'fail';
		}
	}).addCallback(
		function(e){
			assertEquals(e, 'fail');
			asyncTestCase.continueTesting();
		});
	
	asyncTestCase.waitForAsync();
	d.callback();

};
</script>
</body>
</html>
